### ISSUE TITLE
False positive with `SocketFactory.createSocket()`

### ISSUE THREAD

**Issue by msridhar** [kelloggm/object-construction-checker#381]
test case:
```java
  @MustCall("closeSocket")
  class SocketField {
    protected @Owning Socket socket = null;

    @CreatesObligation("this")
    protected void setupConnection(javax.net.SocketFactory socketFactory) throws IOException {
      this.socket.close();
      //this.socket = new Socket(); // no error, code verifies
      this.socket = socketFactory.createSocket(); // error on this line
    }
    @EnsuresCalledMethods(value = "this.socket", methods = "close")
    private void closeSocket() throws IOException {
      this.socket.close();
    }
  }
```
With this test we get a false positive `required.method.not.called` error at the assignment to `this.socket` inside `setupConnection()`.  If I comment that line and uncomment `this.socket = new Socket()` there is no error.  (The socket is never bound in this example, but I don't think that is relevant.)  Looking at `SocketAccumulationFrames.astub` I didn't see an obvious problem:

https://github.com/kelloggm/object-construction-checker/blob/always-call-checker/must-call-checker/src/main/java/org/checkerframework/checker/mustcall/SocketAccumulationFrames.astub

Not sure where to dig deeper.

**Comment by kelloggm** [comment]
@msridhar I've looked into this more. This bug is really insidious, and I'm not confident that I'll be able to produce a solution we can be happy with. What follows are my notes, and some ideas on how to work around this in real code.

The key problem here is that the Called Methods Checker is treating `socketFactory.createSocket()` as a possibly-impure operation. Because it does so, it clears information from its store, such as the fact that `close` has been called on the field `this.socket`. A short-term workaround is to extract the call to `createSocket` into its own method, which can be marked as `@Pure`:

```
    @Pure
    private Socket getSocket(javax.net.SocketFactory socketFactory) throws IOException {
        return socketFactory.createSocket();
    }
```

If the call to `socketFactory.createSocket()` in the code above is replaced with a call to this method, the code will typecheck.

Obviously this is unsatisfying: why does this call need to be wrapped in this way, but not e.g. the `Socket` constructor (since `new Socket()`) works just as well? I don't really know. I also tried adding `@Pure` to the stub file you referred to above, which also doesn't work (why? I also don't know.). 

I also don't know why the store is cleared at all before the method is called. Calling `getStoreBefore(Node)` on the whole assignment node, for example, returns a store with no information about `this.socket` when the RHS is `socketFactory.createSocket()`, even though as the **store before** that operation shouldn't have executed yet. You can see this weirdness by e.g. inserting a local variable; the assignment to `s` below typechecks in the body of `setupConnection`, even though we issue an error on the next line:

```
    @CreatesObligation("this")
    protected void setupConnection(javax.net.SocketFactory socketFactory) throws IOException {
        this.socket.close();
        @CalledMethods("close") Socket s = this.socket;
        this.socket = socketFactory.createSocket();
    }
```
Unfortunately, I can't think of a work-around that exploits this particular quirk of the system.

This does, however, lead me to another work-around for the problem which does avoid issuing the error. Consider the following replacement for the body of `setupConnection`:

```
    @CreatesObligation("this")
    protected void setupConnection(javax.net.SocketFactory socketFactory) throws IOException {
        Socket s = socketFactory.createSocket();
        this.socket.close();
        this.socket = s;
    }
```

This code typechecks, because the call to `close()` comes after the call to `createSocket()`. This might be the easiest way to get around the problem in practice.

Ideally, we'd be able to get the store in a real "before" state in order to do this check, making these workarounds unnecessary. I haven't found a way to produce such a store, however, from the `Node` that's available when we do the check. I also tried to check the store before the block containing the assignment (using direct access to the `flowResult` in the type factory), but that store has similarly already had its value for `this.socket` cleared. That its possible to write an assignment directly before the relevant line that typechecks suggests to me that this should be possible **somehow**, but I don't have any other ideas on how. The only idea I have is that it might be necessary to move the check earlier (i.e. out of the CFG-post-processing stage and into one of the typechecking stages), but I don't think that's feasible since it needs to compare the results of both typecheckers (and it's logically part of the consistency checker).

What do you think? Should we just live with these workarounds? Should I engage Mike, Suzanne, or both on whether it might be possible to get access to a store before the purity check happens?

**Comment by msridhar** [comment]
> I also don't know why the store is cleared at all before the method is called. Calling `getStoreBefore(Node)` on the whole assignment node, for example, returns a store with no information about `this.socket` when the RHS is `socketFactory.createSocket()`, even though as the **store before** that operation shouldn't have executed yet. You can see this weirdness by e.g. inserting a local variable; the assignment to `s` below typechecks in the body of `setupConnection`, even though we issue an error on the next line:
> 
> ```
>     @CreatesObligation("this")
>     protected void setupConnection(javax.net.SocketFactory socketFactory) throws IOException {
>         this.socket.close();
>         @CalledMethods("close") Socket s = this.socket;
>         this.socket = socketFactory.createSocket();
>     }
> ```

This seems to me like a clear bug.  Is there any way we can debug to figure out why the store _before_ the assignment is getting information cleared out of it?  Or is that too nasty?

**Comment by kelloggm** [comment]
> This seems to me like a clear bug

I agree that the behavior is undesirable, but it doesn't violate the contract of the store, which is permitted to "forget" information whenever convenient. I don't know why the store behaves this way; further investigation would have to look into the core CF rather than our checker. While that will likely be nasty, my bigger concern is that this behavior exists for a reason (i.e. some type system depends on it for soundness), so even if we discover the cause it may not be possible to fix.

**Comment by kelloggm** [comment]
fixed by #389
